### PR TITLE
FEAT: Support rerank models

### PR DESCRIPTION
**Pull request by aresnow1**
- [x] Add rerank type
- [x] Compatible with cohere rerank APIs
- [x] Support bge-reranker-large and bge-reranker-base
- [x] Support downloading from modelscope
- [x] Documentation

example for launching rerank model:
```Python
model_uid = client.launch_model(
        model_name="bge-reranker-large", model_type="rerank"
    )
model = client.get_model(model_uid)
query = "A man is eating pasta."
corpus = [
    "A man is eating food.",
    "A man is eating a piece of bread.",
    "The girl is carrying a baby.",
    "A man is riding a horse.",
    "A woman is playing violin.",
    "Two men pushed carts through the woods.",
    "A man is riding a white horse on an enclosed ground.",
    "A monkey is playing drums.",
    "A cheetah is running behind its prey.",
]
print(model.rerank(corpus, query))
```

### CODE REVIEW

File: xinference/model/rerank/__init__.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import json
+import os
+
+from .core import RerankModelSpec, get_cache_status
+
+_model_spec_json = os.path.join(os.path.dirname(__file__), "model_spec.json")
+_model_spec_modelscope_json = os.path.join(
+    os.path.dirname(__file__), "model_spec_modelscope.json"
+)
+BUILTIN_RERANK_MODELS = dict(
```
Comment (ChengjieLi28) [inline xinference/model/rerank/__init__.py:25]: Should we add this model_type to the interface `get_model_registration` and `list_model_registrations`?

File: xinference/model/rerank/core.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def get_cache_status(
+    model_spec: RerankModelSpec,
+) -> bool:
+    cache_dir = os.path.realpath(
+        os.path.join(XINFERENCE_CACHE_DIR, model_spec.model_name)
+    )
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir, exist_ok=True)
```
Comment (codingl2k1) [inline xinference/model/rerank/core.py:133]: I noticed that the `get_cache_status` should not create an empty cache dir. I will create a PR to fix the `get_cache_status` in other model types, could you remove this makedirs in this PR?

File: xinference/model/rerank/__init__.py (hunk 1 above)
Comment (aresnow1) [reply xinference/model/rerank/__init__.py:25]: Done.

File: xinference/model/rerank/core.py (hunk 2 above)
Comment (aresnow1) [reply xinference/model/rerank/core.py]: Done.

**Review by codingl2k1 — APPROVED** [review]
LGTM

**Pull request merged**